Add a new FuzzyPP component.

GlassAndOneHalf opened the pull request:
This commit adds a new "fuzzy preprocessor" which allows users to preprocess their source files, preserving comments, before passing them through to the `fuzzyc2cpg` system.

GlassAndOneHalf commented:
All tests are passing locally - I just need to fix the submodules w/ Travis.

ltcmelo commented:
@GlassAndOneHalf, I'm glad to review this PR. But before doing so, I'd like to understand whether we really need to maitain our own "C preprocessor project" - instead of invoking the user's PP.

Based on a brief Slack chat, you mention the matters of (1) keeping comments and (2) missing headers:

(1) Typically, preprocessors allow you to do that. In Clang/GCC's, this is done with flag `-CC`.
(2) Indeed a problem, but, for certain cases (I believe ours), it should be enough to stub the `.h` in question with an empty file.

fabsx00 commented:
@ltcmelo please trust that we have thought this over and just review the code.

AlexDenisov commented:
@GlassAndOneHalf currently, you are adding `cxxopts` as a submodule, but committing things like `catch2` into the repo. Would it make sense to also move those third-party things into submodules? This will simplify code reviews, otherwise, it is quite hard to follow.

ltcmelo commented:
OK @fabsx00. I'll wait until the the "reorg" of the code base, as @AlexDenisov mentions. But, instead of submodules, my suggestion would be use the libs directly.

GlassAndOneHalf commented:
@AlexDenisov @ltcmelo I've moved `catch2` & `cxxopts` into submodules. `simplecpp` is included directly as I have modified the source slightly in order to preserve comments. You can see the README in the `external/simplecpp` directory to see the change that I made to the source for that project.

AlexDenisov commented:
@GlassAndOneHalf do you think it worth the effort to use a fork of `simplecpp` with your modifications and also use it as a submodule? I think it might be a cleaner solution.

AlexDenisov commented on fuzzypp/CMakeLists.txt:
```diff
@@ -0,0 +1,56 @@
+
+# Building with Clang is not currently supported.
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compile_options(-Weverything
```
I think this is a bit too much to handle. IIRC some warnings there even contradict the others.

GlassAndOneHalf replied:
I need to play around with the Clang support. I don't think it compiles at the moment (not even due to having `-Weverything` turned on!)

AlexDenisov commented on fuzzypp/README.md:
```diff
@@ -0,0 +1,61 @@
+
+FAQ
+---
+Coming soon!
```
Nice readme!

AlexDenisov commented on fuzzypp/src/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+
+target_include_directories(${FUZZYPP_LIB_NAME} PUBLIC
+                           "${CMAKE_SOURCE_DIR}/external/cxxopts/include"
+                           "${CMAKE_SOURCE_DIR}/external/simplecpp")
```
Just a comment, no action needed.

There is an option (`SYSTEM`) to mark header search paths as system headers, this way we won't get any warnings that are coming from third-party libraries. Whether it is desired or not is a separate question though 😄

AlexDenisov commented on fuzzypp/src/cliopts.hpp:
```diff
@@ -0,0 +1,51 @@
+            const std::vector<std::string> include_paths;
+            const std::vector<std::string> defines;
+            const std::vector<std::string> undefines;
+            const std::string output_directory;
```
I assume this is required by the `cxxopts` library?

GlassAndOneHalf replied:
That's right, I parse the options using `cxxopts` and put the results into this data structure.

AlexDenisov commented on fuzzypp/test/cliopts_test.cpp:
```diff
@@ -0,0 +1,258 @@
+    auto result = opts.validate_options();
+    REQUIRE(result.has_value());
+    REQUIRE(*result == "Include paths must not contain '..'.");
+}
```
Hats off, very nice test suite 👍

AlexDenisov commented on fuzzypp/test/main.cpp:
```diff
@@ -0,0 +1,4 @@
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+
+TEST_CASE("1: Test entrypoint", "main") {}
```
I've never used catch before, so I'm curious what this test does?

GlassAndOneHalf replied:
Catch will automatically generate a `main` method for us (see L1). I believe we have to add a dummy test here to get everything to work for reasons unknown. :smile:

AlexDenisov replied:
👍

GlassAndOneHalf commented:
Going to hold off merging this until I can get a SL fork of simplecpp.

ltcmelo commented on fuzzypp/src/preprocessor.hpp:
```diff
@@ -0,0 +1,24 @@
+#include "cliopts.hpp"
+
+namespace fuzzypp::preprocessor {
+    class FuzzyPreprocessor {
```
Silly note, but the preprocessor itself isn't fuzzy ;). This name can be misleading, perhaps just `Preprocessor` or `Fuzzyc2cpgPreprocessor`, or `Fuzzyc2cpgPP`?

ltcmelo commented on fuzzypp/src/preprocessor.cpp:
```diff
@@ -0,0 +1,60 @@
+    FuzzyPreprocessor::generate_simplecpp_opts(const fuzzypp::cliopts::CliOptions& options) {
+        simplecpp::DUI simple_opts;
+
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
```
Why not making the containers in `CliOptions` an `std::list` too, so that you can copy (or even move) them here?

GlassAndOneHalf replied:
Just tried this out. Unfortunately it seems `cxxopts` doesn't support parsing cmd args as a list :(

ltcmelo commented on fuzzypp/test/preprocessor_test.cpp:
```diff
@@ -0,0 +1,332 @@
+#include <catch.hpp>
```
General question concerning behaviour and tests: what's our expecation for macros? For instance, would fuzzyc2cpg interpret function- and object-like macros as is? If they must be expanded, I'd assume that **simplecpp** itself has tests for those, right? Does it handle [self referential macros](https://gcc.gnu.org/onlinedocs/gcc-3.3.6/cpp/Self_002dReferential-Macros.html) well?

GlassAndOneHalf replied:
Yes, I believe it already handles these cases. It won't stop you from infinitely recursing, though :wink:

ltcmelo commented on fuzzypp/src/preprocessor.cpp:
```diff
@@ -0,0 +1,60 @@
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
+        simple_opts.includePaths = std::list<std::string> { options.include_paths.cbegin(), options.include_paths.cend() };
+        simple_opts.defines = std::list<std::string> { options.defines.cbegin(), options.defines.cend() };
+        simple_opts.undefined = std::set<std::string> { options.undefines.cbegin(), options.undefines.cend() };
```
`undefined` x `undefines`

GlassAndOneHalf replied:
As with the `includePaths` - stuck with this name :(

GlassAndOneHalf commented on fuzzypp/external/simplecpp/README.md:
```diff
@@ -0,0 +1,17 @@
+This is a version of [simplecpp](https://github.com/danmar/simplecpp), modified to retain comments in the preprocessor output.
```
Yeah I think this would be the idea long term plan - I just wanted to make a quick & dirty change here :wink:

GlassAndOneHalf commented on fuzzypp/src/cliopts.cpp:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    inline const std::vector<std::string> 
```
Are you sure that's correct? I can't find anything stating this.

https://en.cppreference.com/w/cpp/language/inline
https://isocpp.org/wiki/faq/inline-functions#where-to-put-inline-keyword

ltcmelo replied:
Yes, how would the compiler _inline_ the function if it can't see it? The compiler can't look across different .cpp, only in .h.  Note that putting an inline function **outside the class body** doesn't mean that it has to be in the .cpp, you can (and should) have those definitions in the .h. Check that same page you linked (see screenshot):

<img width="1077" alt="Screen Shot 2019-10-11 at 11 09 45" src="https://user-images.githubusercontent.com/2905588/66658621-39868c00-ec18-11e9-98e5-2d0f96f88820.png">

GlassAndOneHalf commented on fuzzypp/src/cliopts.cpp:
```diff
@@ -0,0 +1,112 @@
+        for (const auto& file : files) {
+            std::filesystem::path p { file };
+            if (!is_path_valid(p)) {
+                return std::optional<std::string> { "File paths must not contain '..'." };
```
When we generate the output, we try to mimic the existing structure of the source code in the filesystem. If the user passes the program a file path with `..`, it can play havoc with this.

GlassAndOneHalf commented on fuzzypp/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+add_compile_options(-O3)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+    add_compile_options(-Wall
```
This generates a _lot_ of errors from `cxxopts`. Will leave this out for the time being.

GlassAndOneHalf commented on fuzzypp/src/cliopts.hpp:
```diff
@@ -0,0 +1,51 @@
+            CliOptions(CliOptions&&) = default;
+
+            static const std::optional<CliOptions> 
+            parse_command_line(int argc, char* argv[]);
```
We produce a library from the source, so clients are able to construct their own `CliOptions` without having to fake cmd line args.

GlassAndOneHalf commented on fuzzypp/src/preprocessor.cpp:
```diff
@@ -0,0 +1,60 @@
+        auto opts = generate_simplecpp_opts(options);
+        const auto output_path = std::filesystem::path { options.output_directory };
+
+        auto write_to_file = [&](const auto& filename) {
```
Nice catch!

GlassAndOneHalf commented on fuzzypp/src/preprocessor.cpp:
```diff
@@ -0,0 +1,60 @@
+        simplecpp::DUI simple_opts;
+
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
+        simple_opts.includePaths = std::list<std::string> { options.include_paths.cbegin(), options.include_paths.cend() };
```
The left hand side of the expression is from a third party library :(

GlassAndOneHalf commented on fuzzypp/src/preprocessor.cpp:
```diff
@@ -0,0 +1,60 @@
+        simplecpp::OutputList output_list;
+        std::vector<std::string> files;
+        
+        std::ifstream input_file { filename };
```
We check them when performing validation on the options in `cliopts.cpp`. Perhaps it would be a good idea to do this anyway? What do you think?

ltcmelo replied:
Generally yes. Even if the path is valid, you'll only know whether you can actually open the file through the stream.
